In [20]:
import tensorflow as tf
from tensorflow.keras.layers import Dense, BatchNormalization, Flatten, Conv2D, Dropout, MaxPool2D
from tensorflow.keras.models import load_model
(x_train,y_train),(x_test,y_test) = tf.keras.datasets.mnist.load_data()
import matplotlib.pyplot as plt

print(tf.__version__)
import numpy as np

2.1.0


# Data Preprocessing

In [21]:
# Reshaping the array to 4-dims so that it can work with the Keras API
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
input_shape = (28, 28, 1)
# Making sure that the values are float so that we can get decimal points after division
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
# Normalizing the RGB codes by dividing it to the max RGB value.
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print('Number of images in x_train', x_train.shape[0])
print('Number of images in x_test', x_test.shape[0])

x_train shape: (60000, 28, 28, 1)
Number of images in x_train 60000
Number of images in x_test 10000


# Finding Appropriate number of Conv Layers

### A) Input -> Conv2d(24,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [26]:
model = tf.keras.models.Sequential()

model.add(Conv2D(24,kernel_size=5,activation='relu',padding='same',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [27]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 5s 88us/sample - loss: 0.1321 - accuracy: 0.9601 - val_loss: 0.0471 - val_accuracy: 0.9841
Epoch 2/20
60000/60000 [==============================] - 5s 85us/sample - loss: 0.0461 - accuracy: 0.9862 - val_loss: 0.0454 - val_accuracy: 0.9845
Epoch 3/20
60000/60000 [==============================] - 5s 85us/sample - loss: 0.0312 - accuracy: 0.9899 - val_loss: 0.0424 - val_accuracy: 0.9862
Epoch 4/20
60000/60000 [==============================] - 5s 85us/sample - loss: 0.0198 - accuracy: 0.9938 - val_loss: 0.0390 - val_accuracy: 0.9874
Epoch 5/20
60000/60000 [==============================] - 5s 88us/sample - loss: 0.0174 - accuracy: 0.9943 - val_loss: 0.0336 - val_accuracy: 0.9901
Epoch 6/20
60000/60000 [==============================] - 5s 90us/sample - loss: 0.0121 - accuracy: 0.9958 - val_loss: 0.0413 - val_accuracy: 0.9875
Epoch 7/20
60000/60000 [==============================] 

### B) Input -> Conv2d(24,5) -> Pool(2) -> Conv2d(48,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [28]:
model = tf.keras.models.Sequential()

model.add(Conv2D(24,kernel_size=5,activation='relu',padding='same',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(48,kernel_size=5,activation='relu',padding='same'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [29]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 9s 146us/sample - loss: 0.1125 - accuracy: 0.9652 - val_loss: 0.0434 - val_accuracy: 0.9858
Epoch 2/20
60000/60000 [==============================] - 8s 134us/sample - loss: 0.0387 - accuracy: 0.9881 - val_loss: 0.0323 - val_accuracy: 0.9887
Epoch 3/20
60000/60000 [==============================] - 8s 135us/sample - loss: 0.0251 - accuracy: 0.9918 - val_loss: 0.0347 - val_accuracy: 0.9882
Epoch 4/20
60000/60000 [==============================] - 8s 135us/sample - loss: 0.0183 - accuracy: 0.9942 - val_loss: 0.0208 - val_accuracy: 0.9929
Epoch 5/20
60000/60000 [==============================] - 8s 134us/sample - loss: 0.0155 - accuracy: 0.9951 - val_loss: 0.0393 - val_accuracy: 0.9895
Epoch 6/20
60000/60000 [==============================] - 8s 137us/sample - loss: 0.0119 - accuracy: 0.9962 - val_loss: 0.0326 - val_accuracy: 0.9918
Epoch 7/20
60000/60000 [==========================

### C) Input -> Conv2d(24,5) -> Pool(2) -> Conv2d(48,5) -> Pool(2) -> Conv2d(64,5) -> Pool(2) ->Flatten -> Dense(128) -> Dense(10)

In [30]:
model = tf.keras.models.Sequential()

model.add(Conv2D(24,kernel_size=5,activation='relu',padding='same',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(48,kernel_size=5,activation='relu',padding='same'))
model.add(MaxPool2D())

model.add(Conv2D(64,kernel_size=5,activation='relu',padding='same'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [31]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 11s 181us/sample - loss: 0.1257 - accuracy: 0.9609 - val_loss: 0.0356 - val_accuracy: 0.9888
Epoch 2/20
60000/60000 [==============================] - 10s 174us/sample - loss: 0.0404 - accuracy: 0.9872 - val_loss: 0.0530 - val_accuracy: 0.9844
Epoch 3/20
60000/60000 [==============================] - 10s 169us/sample - loss: 0.0286 - accuracy: 0.9908 - val_loss: 0.0470 - val_accuracy: 0.9839
Epoch 4/20
60000/60000 [==============================] - 11s 178us/sample - loss: 0.0221 - accuracy: 0.9931 - val_loss: 0.0275 - val_accuracy: 0.9922
Epoch 5/20
60000/60000 [==============================] - 10s 172us/sample - loss: 0.0167 - accuracy: 0.9949 - val_loss: 0.0291 - val_accuracy: 0.9928
Epoch 6/20
60000/60000 [==============================] - 10s 169us/sample - loss: 0.0144 - accuracy: 0.9952 - val_loss: 0.0315 - val_accuracy: 0.9897
Epoch 7/20
60000/60000 [====================

### Conclusion: Although it looks C can leverage better accuracy upon more training the ideal choice seems to be B as it is equally good and also computationally less expensive than C

# Finding appropriate feature map size

### B1) Input -> Conv2d(16,5) -> Pool(2) -> Conv2d(32,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [40]:
model = tf.keras.models.Sequential()

model.add(Conv2D(16,kernel_size=5,activation='relu',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(32,kernel_size=5,activation='relu'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [41]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 5s 86us/sample - loss: 0.1452 - accuracy: 0.9555 - val_loss: 0.0668 - val_accuracy: 0.9777
Epoch 2/20
60000/60000 [==============================] - 5s 79us/sample - loss: 0.0459 - accuracy: 0.9858 - val_loss: 0.0423 - val_accuracy: 0.9859
Epoch 3/20
60000/60000 [==============================] - 5s 83us/sample - loss: 0.0312 - accuracy: 0.9904 - val_loss: 0.0297 - val_accuracy: 0.9892
Epoch 4/20
60000/60000 [==============================] - 5s 84us/sample - loss: 0.0246 - accuracy: 0.9921 - val_loss: 0.0327 - val_accuracy: 0.9900
Epoch 5/20
60000/60000 [==============================] - 5s 85us/sample - loss: 0.0197 - accuracy: 0.9937 - val_loss: 0.0362 - val_accuracy: 0.9886
Epoch 6/20
60000/60000 [==============================] - 5s 83us/sample - loss: 0.0152 - accuracy: 0.9951 - val_loss: 0.0305 - val_accuracy: 0.9917
Epoch 7/20
60000/60000 [==============================] 

### B2) Input -> Conv2d(24,5) -> Pool(2) -> Conv2d(48,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [42]:
model = tf.keras.models.Sequential()

model.add(Conv2D(24,kernel_size=5,activation='relu',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(48,kernel_size=5,activation='relu'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [43]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 6s 100us/sample - loss: 0.1271 - accuracy: 0.9604 - val_loss: 0.0511 - val_accuracy: 0.9842
Epoch 2/20
60000/60000 [==============================] - 6s 95us/sample - loss: 0.0433 - accuracy: 0.9870 - val_loss: 0.0355 - val_accuracy: 0.9875
Epoch 3/20
60000/60000 [==============================] - 6s 96us/sample - loss: 0.0297 - accuracy: 0.9905 - val_loss: 0.0303 - val_accuracy: 0.9901
Epoch 4/20
60000/60000 [==============================] - 6s 94us/sample - loss: 0.0213 - accuracy: 0.9930 - val_loss: 0.0276 - val_accuracy: 0.9922
Epoch 5/20
60000/60000 [==============================] - 6s 96us/sample - loss: 0.0164 - accuracy: 0.9946 - val_loss: 0.0334 - val_accuracy: 0.9897
Epoch 6/20
60000/60000 [==============================] - 6s 96us/sample - loss: 0.0139 - accuracy: 0.9956 - val_loss: 0.0290 - val_accuracy: 0.9912
Epoch 7/20
60000/60000 [==============================]

### B3) Input -> Conv2d(32,5) -> Pool(2) -> Conv2d(64,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [44]:
model = tf.keras.models.Sequential()

model.add(Conv2D(32,kernel_size=5,activation='relu',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(64,kernel_size=5,activation='relu'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [45]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 6s 107us/sample - loss: 0.1190 - accuracy: 0.9645 - val_loss: 0.0505 - val_accuracy: 0.9841
Epoch 2/20
60000/60000 [==============================] - 6s 99us/sample - loss: 0.0389 - accuracy: 0.9877 - val_loss: 0.0314 - val_accuracy: 0.9904
Epoch 3/20
60000/60000 [==============================] - 6s 99us/sample - loss: 0.0265 - accuracy: 0.9918 - val_loss: 0.0374 - val_accuracy: 0.9880
Epoch 4/20
60000/60000 [==============================] - 6s 100us/sample - loss: 0.0193 - accuracy: 0.9941 - val_loss: 0.0245 - val_accuracy: 0.9923
Epoch 5/20
60000/60000 [==============================] - 7s 109us/sample - loss: 0.0165 - accuracy: 0.9952 - val_loss: 0.0333 - val_accuracy: 0.9908
Epoch 6/20
60000/60000 [==============================] - 6s 108us/sample - loss: 0.0120 - accuracy: 0.9961 - val_loss: 0.0343 - val_accuracy: 0.9905
Epoch 7/20
60000/60000 [============================

### B4) Input -> Conv2d(48,5) -> Pool(2) -> Conv2d(96,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [46]:
model = tf.keras.models.Sequential()

model.add(Conv2D(48,kernel_size=5,activation='relu',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(96,kernel_size=5,activation='relu'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [47]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 9s 147us/sample - loss: 0.1147 - accuracy: 0.9649 - val_loss: 0.0401 - val_accuracy: 0.9873
Epoch 2/20
60000/60000 [==============================] - 9s 142us/sample - loss: 0.0366 - accuracy: 0.9887 - val_loss: 0.0350 - val_accuracy: 0.9890
Epoch 3/20
60000/60000 [==============================] - 9s 142us/sample - loss: 0.0257 - accuracy: 0.9914 - val_loss: 0.0286 - val_accuracy: 0.9909
Epoch 4/20
60000/60000 [==============================] - 8s 140us/sample - loss: 0.0189 - accuracy: 0.9941 - val_loss: 0.0227 - val_accuracy: 0.9924
Epoch 5/20
60000/60000 [==============================] - 9s 145us/sample - loss: 0.0152 - accuracy: 0.9952 - val_loss: 0.0230 - val_accuracy: 0.9937
Epoch 6/20
60000/60000 [==============================] - 9s 146us/sample - loss: 0.0113 - accuracy: 0.9965 - val_loss: 0.0393 - val_accuracy: 0.9894
Epoch 7/20
60000/60000 [==========================

### B5) Input -> Conv2d(64,5) -> Pool(2) -> Conv2d(128,5) -> Pool(2) -> Flatten -> Dense(128) -> Dense(10)

In [50]:
model = tf.keras.models.Sequential()

model.add(Conv2D(64,kernel_size=5,activation='relu',input_shape=(28,28,1)))
model.add(MaxPool2D())

model.add(Conv2D(96,kernel_size=5,activation='relu'))
model.add(MaxPool2D())

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [51]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20, validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 24s 404us/sample - loss: 0.1116 - accuracy: 0.9647 - val_loss: 0.0398 - val_accuracy: 0.9856
Epoch 2/20
60000/60000 [==============================] - 11s 179us/sample - loss: 0.0379 - accuracy: 0.9881 - val_loss: 0.0338 - val_accuracy: 0.9887
Epoch 3/20
60000/60000 [==============================] - 10s 175us/sample - loss: 0.0257 - accuracy: 0.9921 - val_loss: 0.0300 - val_accuracy: 0.9914
Epoch 4/20
60000/60000 [==============================] - 11s 177us/sample - loss: 0.0185 - accuracy: 0.9941 - val_loss: 0.0249 - val_accuracy: 0.9924
Epoch 5/20
60000/60000 [==============================] - 11s 177us/sample - loss: 0.0144 - accuracy: 0.9954 - val_loss: 0.0272 - val_accuracy: 0.9913
Epoch 6/20
60000/60000 [==============================] - 11s 175us/sample - loss: 0.0113 - accuracy: 0.9962 - val_loss: 0.0293 - val_accuracy: 0.9914
Epoch 7/20
60000/60000 [====================

### Conclusion: Best Training accuracy and Validation accuracy is achieved in B3

# Adding Dropout

In [56]:
for i in range(1,4):
    dropout = i*0.2
    model = tf.keras.models.Sequential()

    model.add(Conv2D(32,kernel_size=5,activation='relu',input_shape=(28,28,1)))
    model.add(MaxPool2D())
    model.add(Dropout(dropout))
    
    model.add(Conv2D(64,kernel_size=5,activation='relu'))
    model.add(MaxPool2D())
    model.add(Dropout(dropout))

    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(dropout))
    model.add(Dense(10, activation='softmax'))
    
    model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
    model.fit(x=x_train,y=y_train,epochs = 20,verbose=0)
    model.evaluate(x_test,y_test)

10000/10000 [==============================] - 1s 75us/sample - loss: 0.0280 - accuracy: 0.9916


### Conclusion: 40% dropout seems good.

### All these experiments are random and hence I had to run them several times to be sure of the conclusions

# Further improvements

#### 1) Instead of using one 5\*5 kernel in Convolution, we can use two 3\*3 kernels to mimic the behaviour along with increase in non-linearity  
#### 2) Instead of MaxPooling,we can try a Convolutional layer with a stride of 2 and see if there is any improvement
#### 3) We can add BatchNormalisation to reduce overfitting so that model generalises better.
#### 4) Data augmentaition can also be used but model seems to perform sufficiently good without it.

In [63]:
model = tf.keras.models.Sequential()
dropout = 0.40
model.add(Conv2D(32,kernel_size=3,activation='relu',input_shape=(28,28,1)))
model.add(BatchNormalization())
model.add(Conv2D(32,kernel_size=3,activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(32,kernel_size=5,strides=2,padding='same',activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(dropout))

model.add(Conv2D(64,kernel_size=3,activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(64,kernel_size=3,activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(64,kernel_size=5,strides=2,padding='same',activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(dropout))

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(dropout))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(x=x_train,y=y_train,epochs = 20,validation_data=(x_test,y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 19s 325us/sample - loss: 0.1980 - accuracy: 0.9399 - val_loss: 0.0320 - val_accuracy: 0.9896
Epoch 2/20
60000/60000 [==============================] - 18s 302us/sample - loss: 0.0723 - accuracy: 0.9783 - val_loss: 0.0323 - val_accuracy: 0.9888
Epoch 3/20
60000/60000 [==============================] - 18s 306us/sample - loss: 0.0583 - accuracy: 0.9823 - val_loss: 0.0239 - val_accuracy: 0.9932
Epoch 4/20
60000/60000 [==============================] - 19s 309us/sample - loss: 0.0518 - accuracy: 0.9843 - val_loss: 0.0212 - val_accuracy: 0.9925
Epoch 5/20
60000/60000 [==============================] - 19s 312us/sample - loss: 0.0427 - accuracy: 0.9875 - val_loss: 0.0259 - val_accuracy: 0.9919
Epoch 6/20
60000/60000 [==============================] - 19s 309us/sample - loss: 0.0375 - accuracy: 0.9884 - val_loss: 0.0242 - val_accuracy: 0.9919
Epoch 7/20
60000/60000 [====================

In [64]:
model.save('cnn_model.h5')

### Finally after doing various runs in each experiment, I achieved a training accuracy of around 99.6 and a validation accuracy of around 99.4%. Data Augmentation and using the bigger network might increase this a bit further but this seems to be a good balance of accuracy and computational expense.

In [65]:
model.evaluate(x_test,y_test)

10000/10000 [==============================] - 1s 124us/sample - loss: 0.0191 - accuracy: 0.9945


[0.019054458866282948, 0.9945]